craft_application/grammar.py

```diff
@@ -15,38 +15,61 @@
 # with this program.  If not, see <http://www.gnu.org/licenses/>.
 """Grammar processor."""
 
-from typing import Any
+from typing import Any, cast
 
 from craft_grammar import GrammarProcessor  # type: ignore[import-untyped]
 
-_KEYS = [
-    "source",
+# Values that should return as a single object / list / dict.
+_NON_SCALAR_VALUES = [
```

Comment:
the work is already done here but I strongly feel like we could just inspect the type annotations of a `PartSpec` to figure out if a property is a scalar or not, instead of describing all fields exhaustively here.

Oh well it's already done I suppose. We can change this in the future if we feel is valid (in which case this function could just go in craft-grammar itself)